src/Orleans.Core/Messaging/MessageSerializer.cs

```diff
@@ -116,8 +127,33 @@ internal sealed class MessageSerializer : IMessageSerializer
             }
         }
 
+        private void ReadBodyObject<TInput>(Message message, ref Reader<TInput> reader)
+        {
+            var field = reader.ReadFieldHeader();
+
+            IFieldCodec bodyCodec;
+            if (message.Result == ResponseTypes.Success)
+            {
+                bodyCodec = GetRawResponseCodec(field.FieldType);
+                message.Result = ResponseTypes.None;
+            }
+            else
+            {
+                bodyCodec = _codecProvider.GetCodec(field.FieldType);
+            }
+
+            message.BodyObject = bodyCodec.ReadValue(ref reader, field);
+        }
+
         public (int HeaderLength, int BodyLength) Write<TBufferWriter>(ref TBufferWriter writer, Message message) where TBufferWriter : IBufferWriter<byte>
         {
+            IFieldCodec? bodyCodec = null;
+            if (message.Result is ResponseTypes.None && (message.BodyObject as Response)?.GetSimpleResultType() is { } simpleType)
+            {
+                bodyCodec = GetRawResponseCodec(simpleType);
+                message.Result = ResponseTypes.Success;
```

Comment:
It seems odd & unexpected that the `Write` method is mutating the `Message` parameter

Reply:
It bears explanation in code comments